### PR TITLE
fix(gatsby-remark-images): don’t add links if image is already linked

### LINKED ISSUE #6980: [gatsby-remark-images] Autolinking images overrides other links
## Description

When using `gatsby-remark-images`, if the `linkImagesToOriginal` is set to `true` (which is the default), all images will be linked to their original, even if they're already wrapped in a link.

### Steps to reproduce

- Demo: https://priceless-wing-6269cd.netlify.com/
- Repo: https://github.com/jlengstorf/gatsby-remark-images-link-override-bug

```
# clone the repo
git clone https://github.com/jlengstorf/gatsby-remark-images-link-override-bug
cd gatsby-remark-images-link-override-bug/

# install dependencies
npm i

# start the repo
npm run develop
```

Load `http://localhost:8000/

### Expected result

Images with links should preserve those links.

### Actual result

Links are overridden with a link to the image. (Actually, they're double-wrapped, but the HTML parser drops the outer link.)

### Environment

```
  System:
    OS: macOS High Sierra 10.13.6
    CPU: x64 Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 10.1.0 - ~/.nvm/versions/node/v10.1.0/bin/node
    Yarn: 1.9.2 - /usr/local/bin/yarn
    npm: 5.6.0 - ~/.nvm/versions/node/v10.1.0/bin/npm
  Browsers:
    Chrome: 68.0.3440.84
    Firefox: 60.0.2
    Safari: 11.1.2
  npmPackages:
    gatsby: next => 2.0.0-beta.66 
    gatsby-plugin-react-helmet: next => 3.0.0-beta.4 
    gatsby-plugin-sharp: next => 2.0.0-beta.7 
    gatsby-remark-images: next => 2.0.1-beta.9 
    gatsby-source-filesystem: next => 2.0.1-beta.10 
    gatsby-transformer-remark: next => 2.1.1-beta.5 
```

### PR DESCRIPTION
**Pull request by jlengstorf**
Thanks for the help, @pieh!

fixes #6980

### CODE REVIEW

**Comment by KyleAMathews** [comment]
Deploy preview for *using-postcss-sass* failed.

Built with commit af209f9215d8cd9c5ec2f3fda1cbdae1ec40027d

https://app.netlify.com/sites/using-postcss-sass/deploys/5b63a64073f2cf3a34e74d48

File: packages/gatsby-remark-images/src/index.js [hunk 1]
```diff
@@ -30,11 +31,18 @@ module.exports = (
...
+  let markdownImageNodes = []
 
   // This will also allow the use of html image tags
   const rawHtmlNodes = select(markdownAST, `html`)
 
+  visitWithParents(markdownAST, `image`, (node, ancestors) => {
+    const inLink = ancestors.some(ancestor => ancestor.type === `link`)
+    if (!inLink) {
+      markdownImageNodes.push(node)
+    }
```
Comment (jlengstorf) [inline packages/gatsby-remark-images/src/index.js]: This is broken. We still need to fix the relative path, which I'm pretty sure @pieh told me to do and I just... didn't.

Comment (pieh) [reply packages/gatsby-remark-images/src/index.js]: I think we need to process images in both cases (when it's inside link and when it's not) - difference should be whether we wrap image in link here - https://github.com/jlengstorf/gatsby/blob/af209f9215d8cd9c5ec2f3fda1cbdae1ec40027d/packages/gatsby-remark-images/src/index.js#L175-L188

**Comment by pieh** [comment]
Following trend of nonsense examples we had yesterday: ;)
Do we want to handle cases like:
```
<a href="https://example.org">![Mix #1](./images/gatsby.png)</a>
```
or
```
[<img alt="mix #2" src="./images/gatsby.png"/>](https://example.org)
```
with mix of markdown and html?

I would say this is not in scope for this PR - I think we just need to handle markdown image in markdown link here.

**Comment by jlengstorf** [comment]
Alright — I think I cracked this. Added a check for both `link` and `<a>` in the ancestor tree, and added tests for goofy nesting and mixed markup/markdown. This fixes the problem in my example repo for combos like this:

```
[![Screenshot of video.](./images/gatsby.png)](https://example.org)

<a href="https://google.com">**![test](./images/gatsby.png)**</a>

[**<img src="./images/gatsby.png"/>**](https://gatsbyjs.org)
```

**Pull request merged**